### ISSUE TITLE
Incompatibility between pystan (2.18.1.1.dev0) and arviz indexing

### ISSUE THREAD

**Issue by GWeindel** [arviz-devs/arviz#576]
**Describe the bug**
I guess that due to the recent change to 0 indexing in pystan the arviz and pystan indexing don't match anymore. When reading a pystan fit with az.from_pystan() the correspondence pystan -> arviz is then beta[1] -> beta[0] hence beta[0] -> beta[n] with n being the number of parameters.

**To Reproduce**
Just run a model with pyStan with indexed parameters and read it with az.from_pystan()

**Expected behavior**
 beta[0] -> beta[0]

**Additional context**
arviz 0.3.2
pystan 2.18.1.1.dev0

I guess this should be an easy fix, I can take a look when I have a bit more time !

**Comment by ahartikainen** [comment]
Wait, what? When did PyStan moved to 0 indexing?

**Comment by GWeindel** [comment]
Oh wait I was wrong, I misremembered your announcement on the stan forum. But I guess that the bug occurred when I updated pystan to 2.18, when it was 0-indexed the from_pystan() behaved as expected.

**Comment by ahartikainen** [comment]
What message?

I did change summary to 1 index (from 0) long time ago.

Samples should come from stan, so they are 1-index.

**Comment by GWeindel** [comment]
```
az.summary(ter_fitaz, var_names="beta", round_to=3)

output : 
  | mean | sd | mc error | hpd 3% | hpd 97% | ess | r_hat
-- | -- | -- | -- | -- | -- | -- | --
0.064 | 0.011 | 0.0 | 0.042 | 0.086 | 3004.0 | 1.0
-0.003 | 0.003 | 0.0 | -0.008 | 0.003 | 8447.0 | 1.0
-0.050 | 0.011 | 0.0 | -0.070 | -0.030 | 5102.0 | 1.0
-0.009 | 0.004 | 0.0 | -0.017 | -0.002 | 7529.0 | 1.0
-0.104 | 0.017 | 0.0 | -0.139 | -0.073 | 1731.0 | 1.0

print(fit)

```
(I was refereing to the announcement on discourse https://discourse.mc-stan.org/t/pystan-announcement-one-based-indexing-for-the-names-flatnames/3339)

**Comment by ahartikainen** [comment]
That is super weird. Have I missed something?

So this is not happening with the current release, but dev version only?

(All pystan variables should start with 1)

What does your dictionary say? `fit.sim.samples[0].chains` (or something like that)

**Comment by GWeindel** [comment]
I switched from 2.17.1.0 straight to the dev version, hence I assumed that the 2.18 broke something but I can try the latest stable version.

The output from fit.sim['samples'][0].chains is also 0 indexed
```
OrderedDict([('alpha',
              array([0.7971815 , 0.7971815 , 0.7971815 , ..., 0.41622173, 0.41681073,
                     0.41931551])),
             ('beta[0]',
              array([ 0.63567185,  0.63567185,  0.63567185, ..., -0.09811635,
                     -0.09868827, -0.10114054])),
             ('beta[1]',
              array([1.34902943, 1.34902943, 1.34902943, ..., 0.06776995, 0.07119223,
                     0.07735488])),
             ('beta[2]',
              array([-0.02117017, -0.02117017, -0.02117017, ..., -0.00456911,
                      0.00209434, -0.00642094])),
             ('beta[3]',
              array([-0.28570089, -0.28570089, -0.28570089, ..., -0.06268226,
                     -0.04333432, -0.05396976])),
             ('beta[4]',
              array([-1.10940649, -1.10940649, -1.10940649, ..., -0.00979361,
                     -0.02987596, -0.00726421])),
```

**Comment by ahartikainen** [comment]
Hmm, where is that change coming from.

**Comment by GWeindel** [comment]
I think I got it, the data is fitted on a cluster with the pystan 2.17.1.0 and read with one on the dev version... So I guess this is mixing everything up.

**Comment by ahartikainen** [comment]
Weird, why is it 0-index?

**Comment by GWeindel** [comment]
I fitted the model again with the dev version only and the bug is still present so it is not the mixing between versions...

**Comment by ahartikainen** [comment]
```
import pystan

stan_code = """
data {
    int N;
}
parameters {
    real beta[N];
    vector[N] eta;
}
model {
    beta ~ normal(0,2);
    eta ~ normal(0,3);
}
"""

sm = pystan.StanModel(model_code=stan_code)
stan_data = {"N" : 5}
fit = sm.sampling(data=stan_data)
print(fit)
print(fit.sim["samples"][0].chains)
```

```
Inference for Stan model: anon_model_4d9e1ce55773813f551358db9e3c2f42.
4 chains, each with iter=2000; warmup=1000; thin=1; 
post-warmup draws per chain=1000, total post-warmup draws=4000.

          mean se_mean     sd   2.5%    25%    50%    75%  97.5%  n_eff   Rhat
beta[1] 9.4e-3    0.02   2.08  -4.02  -1.42 5.9e-3   1.43    4.1   7615    1.0
beta[2]-7.8e-4    0.03   1.98  -3.83  -1.38   0.04   1.34   3.78   5905    1.0
beta[3] 4.9e-3    0.02   1.96  -3.85  -1.28   0.02    1.3   3.95   6535    1.0
beta[4]   0.02    0.02    2.0  -3.85  -1.34   0.03   1.41   3.93   7175    1.0
beta[5]-7.9e-3    0.02    2.0  -3.86  -1.39  -0.04   1.37   3.99   6662    1.0
eta[1]    0.05    0.04   3.01  -5.83  -1.99 9.5e-3   2.13   5.99   6186    1.0
eta[2]   -0.04    0.04   2.98  -5.74  -2.04  -0.06   1.97   5.63   6228    1.0
eta[3]    0.05    0.04   2.98  -5.98  -1.88   0.04   2.06   5.97   6394    1.0
eta[4]    0.05    0.04   2.93  -5.72  -1.89   0.05   2.06   5.74   6509    1.0
eta[5]    0.03    0.04   2.95  -5.63  -1.95 9.7e-3   2.04   5.83   6176    1.0
lp__     -4.96    0.05   2.21 -10.03  -6.25  -4.61  -3.38  -1.67   1887    1.0

Samples were drawn using NUTS at Wed Jan 30 21:46:27 2019.
For each parameter, n_eff is a crude measure of effective sample size,
and Rhat is the potential scale reduction factor on split chains (at 
convergence, Rhat=1).
```

```
OrderedDict([
('beta[1]', array([-2.03896867, -2.03896867, -1.32771799, ...,  2.10059995,  -3.24056943,  2.59056912])), 
('beta[2]', array([-0.99486308, -0.99486308,  1.01895926, ...,  1.85701214, -1.51621923,  1.23196716])), 
('beta[3]', array([ 2.01049965,  2.01049965,  2.50466674, ...,  2.52346588, -0.87180193,  0.87321153])), 
('beta[4]', array([-2.4363122 , -2.4363122 , -2.99570026, ..., -1.34848475, -0.27702799,  0.34742366])), 
('beta[5]', array([ 3.26753267,  3.26753267,  3.84159359, ...,  1.97946115, -1.29052829,  1.32850716])), 
('eta[1]', array([-2.88654295, -2.88654295,  5.15546663, ...,  0.87958961, 2.13888426, -2.27803499])), 
('eta[2]', array([ 2.16330293,  2.16330293, -0.93743613, ...,  2.01474743, -0.22275823,  0.96590145])), 
('eta[3]', array([ 0.93716777,  0.93716777,  1.42937942, ...,  0.45988202, -3.68913017,  3.68166533])), 
('eta[4]', array([ 8.80459379,  8.80459379, -5.96187178, ...,  8.37544423, -4.14646902,  4.44352735])), 
('eta[5]', array([ 0.24038022,  0.24038022,  1.73294387, ..., -1.92978352, 0.78436099, -0.39551512])), 
('lp__', array([-8.3068134 , -8.3068134 , -7.88124723, ..., -6.87994687, -3.9151714 , -3.55841754]))])
```

Do you have a custom installation? What kind of model do you have?

**Comment by ahartikainen** [comment]
I'm not sure should we handle this or not...

We could add a check that finds the lowest index, and this will work until we have real sparse structures on Stan (this is coming at some point).

**Comment by GWeindel** [comment]
Got it ! When compiling your model in pystan < 2.18 the parameters are 0 indexed. When compiling the model with pystan > 2.18 they are 1 indexed. 
Confusion was generated by the fact that the model was translated to c++ in 2.17 and stored in a pickle file, hence even if I was fitting on 2.18 the c++ code was still from 2.17... and that is the reason why the indexes got messed up... So it is the problem of my custom installation across my computers.

Sorry for this troubling issue

**Comment by ahartikainen** [comment]
No problem. I still need to investigate what have happened between 2.17 and 2.18. Or have you edited the cpp file?

**Comment by ahartikainen** [comment]
Oh, I get it. Did I edit the inner structure?

If so, we need to fix this on ArviZ.

**Comment by GWeindel** [comment]
> Did I edit the inner structure?

What do you mean by that ?

**Comment by ahartikainen** [comment]
The samples are in a dictionary inside fit.sim. And now I'm trying to remember did I change cython code where we subtracted 1 from the Stan parameters or did I edit the Stan summary?

**Comment by GWeindel** [comment]
It works for me ! I leave it up to you to close the issue when you're done testing

**Comment by ahartikainen** [comment]
Closed by #578
